### PR TITLE
Physi cs

### PR DESCRIPTION
**Pull request by aletomsic**
This pull request implements the PhysiCS protocol, which can now selected from the antidote config file. 
The changes include:
- making antidote's code more modular for being able to run all these protocols.
- implemented writes and reads in parallel for the new interface (read_objects, update_objects). Before, there were only parallel reads for static transactions, i.e., when running the "clocksi_execute_tx" function of the old interface.

The changes in the code are significant, I can go through the changes with the reviewer to help.

### CODE REVIEW

File: .idea/emacs.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (bieniusa) [inline .idea/emacs.xml]: Please remove!

File: .idea/rebar.xml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (bieniusa) [inline .idea/rebar.xml]: Please remove!Please remove!

File: include/antidote.hrl [hunk 3]
```diff
@@ -50,7 +52,7 @@
 %% wake up and retry. I.e. a read waiting for
 %% a transaction currently in the prepare state that is blocking
 %% that read.
--define(SPIN_WAIT, 10).
+-define(SPIN_WAIT, 0).
```
Comment (bieniusa) [inline include/antidote.hrl]: Why are you setting SPIN_WAIT to 0?

File: include/antidote.hrl [hunk 4]
```diff
@@ -137,9 +142,33 @@
...
+    type :: type(),
+    op_param :: op(),
+    snapshot_vc :: vectorclock(),
+    dependency_vc :: vectorclock(),
+    dc_and_commit_time :: commit_time(),
+    txid :: txid()}).
+
+-record(transaction, {
+    snapshot_clock :: snapshot_time(),
+    transactional_protocol :: atom(),
```
Comment (bieniusa) [inline include/antidote.hrl]: Can you make this type more concrete than just atom()?

File: include/antidote.hrl [hunk 5]
```diff
@@ -137,9 +142,33 @@
...
+    server_pid :: pid(),
+    snapshot_vc :: vectorclock(),
+    txn_id :: txid() | no_txn_inserting_from_log,
+    physics_read_metadata :: physics_read_metadata()}).
+
+
+-record(physics_read_metadata, {
```
Comment (bieniusa) [inline include/antidote.hrl]: Clean-up commented code?

File: include/antidote.hrl [hunk 6]
```diff
@@ -164,11 +193,24 @@
 -type log_id() :: [partition_id()].
 -type bucket() :: term().
 -type snapshot() :: term().
+-type orddict() :: orddict().
```
Comment (bieniusa) [inline include/antidote.hrl]: ??? Seems superfluous

File: include/antidote.hrl [hunk 7]
```diff
@@ -201,27 +243,34 @@
...
+    commit_protocol :: term(),
+    state :: active | prepared | committing | committed | undefined
+    | aborted | committed_read_only,
+    operations :: undefined | list(),
+    internal_read_set :: orddict(),
+    return_accumulator :: list() | ok | {error, reason()},
+    is_static :: boolean(),
+    full_commit :: boolean(),
+    stay_alive :: boolean(),
+    %% The following are needed by the physics protocol
```
Comment (bieniusa) [inline include/antidote.hrl]: Clean-up!

File: rebar.config [hunk 8]
```diff
@@ -9,8 +9,8 @@
     {erlzmq, {git, "git://github.com/tcrain/erlzmq2", {branch, "master"}}},
     %% antidote_pb is client interface. Needed only for riak_tests.
     {antidote_pb, {git, "git://github.com/syncfree/antidote_pb", {branch, "new_interface_erl19"}}},
-    {antidote_crdt, ".*", {git, "git://github.com/syncfree/antidote_crdt", {tag, "0.0.2"}}},
-    {antidote_utils, {git, "git://github.com/syncfree/antidote_utils", {tag, "0.0.1"}}},
+    {antidote_crdt, ".*", {git, "git://github.com/syncfree/antidote_crdt", {branch, "master"}}},
```
Comment (bieniusa) [inline rebar.config]: Make tags in these repos!
We need repeatable builds.

File: test/multiple_dcs_SUITE.erl [hunk 9]
```diff
@@ -66,7 +66,7 @@ init_per_testcase(_Case, Config) ->
 end_per_testcase(_, _) ->
```
Comment (bieniusa) [inline test/multiple_dcs_SUITE.erl]: Clean up in this file!
Remove comments and check whitespace, such that only minimal changes are made.
Also, please check this for *all* other files.

File: src/antidote.app.src [hunk 10]
```diff
@@ -25,6 +25,7 @@
   %% txn_prot:
   %%        clocksi -> uses "Cure" protocol to define snapshots and causal dependencies (https://pages.lip6.fr/Marc.Shapiro/papers/Cure-final-ICDCS16.pdf)
   %%        gr -> uses "Gentle-rain like" protocol to define snapshots and causal dependencies (https://infoscience.epfl.ch/record/202079)
+  %%        physics -> uses physiCS like protocol to define snapshots, causal dependencies and commitment protocol (https://pages.lip6.fr/Marc.Shapiro/papers/PhysiCS-NMSI-PaPoC-2016.pdf)
```
Comment (bieniusa) [inline src/antidote.app.src]: Add documentation for this option.

**Comment by tcrain** [comment]
My main suggestion is to be careful for duplicated code where you do the "if protocol type".  It seems that in many of these cases a lot of the code is duplicated, and sometimes it is just because of a minor structural difference.  Maybe try to unify the structures, or use new ones.  In the places where there are big differences it might help to separate out the code into different methods and add some comments about the differences to help the readability.
https://github.com/SyncFree/antidote/pull/223/files?diff=split#diff-cf224960a5452d2886845b108ffa2a14R349
https://github.com/SyncFree/antidote/pull/223/files?diff=split#diff-4389c408eafb563cc5be6435fa123ad1R644


Maybe make a define for this line since it is used a lot
Protocol when ((Protocol == gr) or (Protocol == clocksi))


The perform single item update was removed.  I guess this is ok since it wasn't very clean, but maybe it is worth creating an issue where you describe what is missing in terms of "non-transactional" operations and how/where they could be implemented cleanly.


There are some places where you removed comments about what the code was doing, I know the commenting isn't good as it is, but as we go we should try to improve them:
https://github.com/SyncFree/antidote/pull/223/files?diff=split#diff-478fd85134cec43d44f0115a26d7a118L351

There are several removed tests:
https://github.com/SyncFree/antidote/pull/223/files?diff=split#diff-2afce300e6253387a488fb922191614dL71
https://github.com/SyncFree/antidote/pull/223/files?diff=split#diff-bf06b660c0c38ec49b7df377c5ad545bL37
https://github.com/SyncFree/antidote/pull/223/files?diff=split#diff-11695bdc58f3891c1f98ae9153b08c9eL70

File: .idea/emacs.xml (hunk 1 above)
Comment (aletomsic) [inline .idea/emacs.xml]: done.

File: .idea/rebar.xml (hunk 2 above)
Comment (aletomsic) [inline .idea/rebar.xml]: done

File: include/antidote.hrl (hunk 3 above)
Comment (aletomsic) [inline include/antidote.hrl]: that's how much time cure waits when there is a transaction being prepared. 
I got, by running benchmarks, that throughput and latency increase when setting it to 0.
should I put it back to 10 anyway?

File: include/antidote.hrl (hunk 4 above)
Comment (aletomsic) [inline include/antidote.hrl]: Defined a type transactional_protocol() in antidote.hrl that can be of physics | clocksi | gr

File: include/antidote.hrl (hunk 6 above)
Comment (aletomsic) [inline include/antidote.hrl]: Removed.

File: rebar.config (hunk 8 above)
Comment (aletomsic) [inline rebar.config]: done.

File: src/antidote.app.src (hunk 10 above)
Comment (aletomsic) [inline src/antidote.app.src]: where should I add it?

**Comment by aletomsic** [comment]
> On Oct 10, 2016, at 05:22, tcrain <notifications@github.com> wrote:
> 
> My main suggestion is to be careful for duplicated code where you do the "if protocol type". It seems that in many of these cases a lot of the code is duplicated, and sometimes it is just because of a minor structural difference. Maybe try to unify the structures, or use new ones. In the places where there are big differences it might help to separate out the code into different methods and add some comments about the differences to help the readability.
> https://github.com/SyncFree/antidote/pull/223/files?diff=split#diff-cf224960a5452d2886845b108ffa2a14R349 <https://github.com/SyncFree/antidote/pull/223/files?diff=split#diff-cf224960a5452d2886845b108ffa2a14R349>
> https://github.com/SyncFree/antidote/pull/223/files?diff=split#diff-4389c408eafb563cc5be6435fa123ad1R644 <https://github.com/SyncFree/antidote/pull/223/files?diff=split#diff-4389c408eafb563cc5be6435fa123ad1R644>sounds good, working on it.
> Maybe make a define for this line since it is used a lot
> Protocol when ((Protocol == gr) or (Protocol == clocksi))
> 
Good idea, but can’t seem to find a way to do it. Do you know how?

File: include/antidote.hrl (hunk 3 above)
Comment (bieniusa) [inline include/antidote.hrl]: Why does setting it to 0 have an effect on the benchmarks?

Comment (bieniusa) [inline include/antidote.hrl]: Why does it have this effect on the benchmark?

File: include/antidote.hrl (hunk 6 above)
Comment (bieniusa) [inline include/antidote.hrl]: Replacing orddict() by list() is not solving the issue I think. There are semantic differences.

**Comment by bieniusa** [comment]
Tests are failing on Travis:
multiple_dcs_SUITE:'-replicated_set_test/1-fun-13-' failed on line 401
Reason: assertEqual

physics_SUITE:'-physics_test_prepare/1-fun-7-' failed on line 323
Reason: assertEqual

File: include/antidote.hrl [hunk 11]
```diff
@@ -15,6 +15,7 @@
 -define(ZMQ_TIMEOUT, 5000).
 -define(NUM_W, 2).
 -define(NUM_R, 2).
+-define(GLOBAL_BUCKET, antidote_bucket).
```
Comment (bieniusa) [inline include/antidote.hrl]: What is this needed for? Comments please!

File: include/antidote.hrl [hunk 12]
```diff
@@ -78,6 +79,9 @@
 %% Frequency at which manager requests remote resources.
 -define(TRANSFER_FREQ, 100). %in Milliseconds
 
+%% threshold time for physics static reads, in microseconds
```
Comment (bieniusa) [inline include/antidote.hrl]: This comment is not descriptive. What kind of threshold is this about??

**Comment by bieniusa** [comment]
Sorry, I can't make this code review. My laptop is totally overloaded with 3.000 lines of changes, it cannot display the review mode.

You have to split this commit in smaller subcommits. 
Alternatively, you can try to convince @cmeiklejohn that this pull request is mergeable.
Not sure what works better for you.

**Comment by bieniusa** [comment]
Dialyzer does not pass:
_build/default/lib/antidote/src/clocksi_interactive_tx_coord_fsm.erl
 127: Function init/1 has no local return
 157: Function start_tx/2 has no local return
 160: Function start_tx_internal/4 has no local return
 167: Function create_transaction_record/6 has no local return
 188: Record construction #transaction{transactional_protocol::'physics',server_pid::'undefined',txn_id::#tx_id{local_start_time::non_neg_integer(),server_pid::pid()},physics_read_metadata::#physics_read_metadata{}} violates the declared type of field server_pid::pid()
 209: Record construction #transaction{snapshot_clock::non_neg_integer(),transactional_protocol::'clocksi' | 'gr',server_pid::'undefined',txn_id::#tx_id{local_start_time::non_neg_integer(),server_pid::pid()},physics_read_metadata::'undefined'} violates the declared type of field server_pid::pid() and physics_read_metadata::#physics_read_metadata{}
 221: Function perform_singleitem_read/2 has no local return

_build/default/lib/antidote/src/clocksi_static_tx_coord_fsm.erl
 101: Function init/1 has no local return
 109: Function start_tx/2 has no local return
 112: Function start_tx_internal/5 has no local return

_build/default/lib/antidote/src/clocksi_vnode.erl
 730: Function filter_updates_per_key_clocksi_test/0 has no local return
 741: Record construction #transaction{transactional_protocol::'clocksi',server_pid::'undefined',txn_id::'undefined',physics_read_metadata::'undefined'} violates the declared type of field server_pid::pid() and txn_id::'no_txn_inserting_from_log' | #tx_id{local_start_time::non_neg_integer(),server_pid::pid()} and physics_read_metadata::#physics_read_metadata{}
 747: Function filter_updates_per_key_physics_test/0 has no local return
 758: Record construction #transaction{transactional_protocol::'clocksi',server_pid::'undefined',txn_id::'undefined',physics_read_metadata::'undefined'} violates the declared type of field server_pid::pid() and txn_id::'no_txn_inserting_from_log' | #tx_id{local_start_time::non_neg_integer(),server_pid::pid()} and physics_read_metadata::#physics_read_metadata{}

_build/default/lib/antidote/src/inter_dc_dep_vnode.erl
 146: Record construction #transaction{server_pid::'undefined',txn_id::'undefined',physics_read_metadata::'undefined'} violates the declared type of field server_pid::pid() and txn_id::'no_txn_inserting_from_log' | #tx_id{local_start_time::non_neg_integer(),server_pid::pid()} and physics_read_metadata::#physics_read_metadata{}
 148: The created fun has no local return
 234: The created fun has no local return
 236: Record construction #operation_payload{type::atom(),op_param::{'merge',_} | {'update',_},dc_and_commit_time::{'undefined' | {atom(),tuple()},non_neg_integer()},txid::'undefined'} violates the declared type of field txid::#tx_id{local_start_time::non_neg_integer(),server_pid::pid()}

_build/default/lib/antidote/src/materializer_vnode.erl
 728: Record construction #transaction{server_pid::'undefined',txn_id::'no_txn_inserting_from_log',physics_read_metadata::'undefined'} violates the declared type of field server_pid::pid() and physics_read_metadata::#physics_read_metadata{}

**Comment by aletomsic** [comment]
fixed.

**Comment by aletomsic** [comment]
> On Oct 10, 2016, at 22:44, Annette Bieniusa <notifications@github.com> wrote:
> 
> Sorry, I can't make this code review. My laptop is totally overloaded with 3.000 lines of changes, it cannot display the review mode.
> 
> You have to split this commit in smaller subcommits. 
> Alternatively, you can try to convince @cmeiklejohn <https://github.com/cmeiklejohn> that this pull request is mergeable.
> Not sure what works better for you.
> 
> 

Hi, I agree the changes are too big. Here is what I propose:

I will make a full pass on the changes to:
	-minimize the changes with master.
	-add docs and comments.
	-create more detailed issues and relate to them in the pr. 

Then, we can arrange a skype to go through the chances together, so I explain your doubts while you review.
I think that would simplify things.

Do you think that would work?

Sorry again for the super large changes.

**Comment by aletomsic** [comment]
added explanation.

**Comment by aletomsic** [comment]
added explanation.

File: include/antidote.hrl (hunk 4 above)
Comment (aletomsic) [inline include/antidote.hrl]: :P
fixed.

**Comment by bieniusa** [comment]
> Am 11.10.2016 um 11:59 schrieb Alejandro Zlatko Tomsic <notifications@github.com>:
> 
> 
> > On Oct 10, 2016, at 22:44, Annette Bieniusa <notifications@github.com> wrote:
> > 
> > Sorry, I can't make this code review. My laptop is totally overloaded with 3.000 lines of changes, it cannot display the review mode.
> > 
> > You have to split this commit in smaller subcommits. 
> > Alternatively, you can try to convince @cmeiklejohn <https://github.com/cmeiklejohn> that this pull request is mergeable.
> > Not sure what works better for you.
> > 
> > 
> 
> Hi, I agree the changes are too big. Here is what I propose:
> 
> I will make a full pass on the changes to:
> -minimize the changes with master.
> -add docs and comments.
> -create more detailed issues and relate to them in the pr. 
> 
> Then, we can arrange a skype to go through the chances together, so I explain your doubts while you review.
> I think that would simplify things.
> 
> Do you think that would work?
> 
> Sorry again for the super large changes.

I cannot inspect the changes because my browser is overloaded when I open the github page which displays the code changes.

Can you please spit issues?
I would like to have one pull request for the renaming that is not related to Physics.
Then, for each other improvement (bug fix) a separate one.
Then one that is cleaning up (white space, comments).
Then one on the Physics protocol.

Besides, you should check the result of Travis - it still fails your branch.
Make sure that the changes you commit do not break the build, also reducing the number of commits might be helpful - you are currently flooding my mailbox, and it is difficult to understand which commits are addressing which issue.

**Comment by cmeiklejohn** [comment]
This pull request is too difficult to merge and we can't do it.

There's a few problems here:

* Your protocol is mixed in with existing improvements to the structure of the application.  That makes it very difficult to evaluate your protocol along with changes to the existing system.
* It's still unclear how to properly test these changes.

I suggest the only way we can merge this is to reapply your changes to a fresh copy of master and break them down into protocol improvements, structural improvements, and introduction of the new PhysiCS protocol.

**Comment by aletomsic** [comment]
Ok, will do that.
Should I “delete” the existing PR?

**Comment by aletomsic** [comment]
I've closed the pull request, will open smaller ones in the future.

**Comment by aletomsic** [comment]
The doc/ folder (that contains the doc/sample_test.erl file) is in the .gitignore file in master.
That’s why in my PR it seems like I removed it, while what happens is that its not being tracked.
I guess we need to keep it (and remove the folder from gitignore, right?

**Comment by bieniusa** [comment]
Does anybody need this sample test? what is it good for?

**Pull request closed without merging**